nixos/modules/services/networking/firezone/gui-client.nix

```
@@ -0,0 +1,142 @@
+{
+  lib,
+  pkgs,
+  config,
+  ...
+}:
+let
+  inherit (lib)
+    boolToString
+    getExe'
+    mkEnableOption
+    mkIf
+    mkOption
+    mkPackageOption
+    types
+    ;
+
+  cfg = config.services.firezone.gui-client;
+in
+{
+  options = {
+    services.firezone.gui-client = {
+      enable = mkEnableOption "the firezone gui client";
+      package = mkPackageOption pkgs "firezone-gui-client" { };
+
+      allowedUsers = mkOption {
+        type = types.listOf types.str;
+        default = [ ];
+        description = ''
+          All listed users will become part of the `firezone-client` group so
+          they can control the IPC service. This is a convenience option.
+        '';
+      };
+
+      name = mkOption {
+        type = types.str;
+        description = "The name of this client as shown in firezone";
+      };
+
+      logLevel = mkOption {
+        type = types.str;
+        default = "info";
+        description = ''
+          The log level for the firezone application. See
+          [RUST_LOG](https://docs.rs/env_logger/latest/env_logger/#enabling-logging)
+          for the format.
+        '';
+      };
+
+      enableTelemetry = mkEnableOption "telemetry";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.groups.firezone-client.members = cfg.allowedUsers;
+
+    # Required for deep-link mimetype registration
+    environment.systemPackages = [ cfg.package ];
+
+    # Required for the token store in the gui application
+    services.gnome.gnome-keyring.enable = true;
+
+    systemd.services.firezone-ipc-service = {
+      description = "GUI IPC service for the Firezone zero-trust access platform";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      path = [ pkgs.util-linux ];
+      script = ''
+        # If FIREZONE_ID is not given by the user, use a persisted (or newly generated) uuid.
+        if [[ -z "''${FIREZONE_ID:-}" ]]; then
+          if [[ ! -e client_id ]]; then
+            uuidgen -r > client_id
+          fi
+          export FIREZONE_ID=$(< client_id)
+        fi
+
+        # run-debug logs to stdout instead of file, which is preferrable
+        exec ${getExe' cfg.package "firezone-client-ipc"} run-debug
```

```suggestion
        exec ${getExe' cfg.package "firezone-client-ipc"} run
```

The `run-debug` command is only meant for debugging. Please don't rely on it for production. We can promise any stability around it whatsoever.

---

The "Export Logs" functionality of the client will be broken if you don't log to files.

Patching the function here wouldn't be too difficult to include stdout logging: https://github.com/firezone/firezone/blob/2d70a8ed31ef0718b65a79eed4cb7a5badeea455/rust/headless-client/src/ipc_service.rs#L661

---

Ah that's too bad. I opted for run-debug specifically because it doesn't log to a file. Since this is a root service, we explicitly want it to log to the system journal via stdout/err to have all service logs centralized. I didn't see an option to do this without run-debug.

---

I agree it would be better if we'd log to stdout for systemd services. Unfortunately, we haven't had the bandwidth yet to implement this. The same codebase is used on Windows as well and we didn't want to have too many diverging codepaths.

The easiest way to achieve this would be to patch the above function to _also_ log to stdout in addition to the files. Then you don't have to rely on `run-debug` which I'd really ask you not to.

I think the ideal solution would be to only log to stdout for systemd services and adapt our "export logs" functionality to use journalctl on Linux.

Unfortunately, Windows sends stdout to /dev/null for Windows Services so we'd always have to log to a file there I think. Open to other solutions as well if anybody knows Windows well.

---

I opened https://github.com/firezone/firezone/issues/8173. Contributions & discussions as always welcome!

---

The latest release now logs to stdout in addition to the files.